File: resource-registry/src/main/java/nl/haploid/octowight/registry/repository/SequenceService.java
package nl.haploid.octowight.registry.repository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.FindAndModifyOptions;
import org.springframework.data.mongodb.core.MongoOperations;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.mongodb.core.query.Update;
import org.springframework.stereotype.Service;

@Service
public class SequenceService {

	@Autowired
	private MongoOperations mongoOperations;

	public long getNextValue(final String key) {
		final Sequence sequence = incrementSequence(key);
		if (sequence != null) {
			return sequence.getValue();
		}
		return startSequence(key).getValue();
	}

	protected Sequence startSequence(final String key) {
		final Sequence sequence = new Sequence();
		sequence.setKey(key);
		sequence.setValue(0l);
		mongoOperations.save(sequence);
		return sequence;
	}

	protected Sequence incrementSequence(final String key) {
		final Query query = new Query(Criteria.where("_id").is(key));
		final Update update = new Update().inc("value", 1);
		final FindAndModifyOptions options = new FindAndModifyOptions().returnNew(true);
		return mongoOperations.findAndModify(query, update, options, Sequence.class);
	}
}

File: resource-registry/src/main/scala/nl/haploid/octowight/registry/repository/SequenceService.scala
package nl.haploid.octowight.registry.repository

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.mongodb.core.query.{Criteria, Query, Update}
import org.springframework.data.mongodb.core.{FindAndModifyOptions, MongoOperations}
import org.springframework.stereotype.Service

@Service
class SequenceService {
  @Autowired private[this] val mongoOperations: MongoOperations = null

  def nextValue(key: String): Long = incrementSequence(key).getOrElse(startSequence(key)).value

  protected def startSequence(key: String): SequenceDmo = {
    val sequenceDmo = new SequenceDmo(key = key, value = 0L)
    mongoOperations.save(sequenceDmo)
    sequenceDmo
  }

  protected def incrementSequence(key: String): Option[SequenceDmo] = {
    val query = new Query(Criteria.where("_id").is(key))
    val update = new Update().inc("value", 1)
    val options = new FindAndModifyOptions().returnNew(true)
    Option(mongoOperations.findAndModify(query, update, options, classOf[SequenceDmo]))
  }
}
